addons/resource/models/resource.py

```diff
@@ -516,7 +516,7 @@ def _leave_intervals_batch(self, start_dt, end_dt, resources=None, domain=None,
         tz_dates = {}
         for leave in self.env['resource.calendar.leaves'].search(domain):
             for resource in resources_list:
-                if leave.resource_id.id not in [False, resource.id]:
+                if leave.resource_id.id not in [False, resource.id] or (not leave.resource_id and resource and resource.company_id.id != leave.company_id.id):
```

Comment:
```suggestion
                if leave.resource_id.id not in [False, resource.id] or (not leave.resource_id and resource and resource.company_id != leave.company_id):
```

Reply:
fixed but why comparing the records instead of just ids ?

Reply:
It is equivalent, when we compare the recordset, the ORM will compare the model name and the ids contained in the both recordsets (see https://github.com/odoo/odoo/blob/45fcb0ac75aebd8bf90465647ee1019095f7356f/odoo/models.py#L6475-L6482).

It is more a good practise because we could have more than one record in the recordsets.